src/System.Linq/src/System/Linq/Concatenate.cs

```diff
@@ -45,5 +32,227 @@ private static IEnumerable<TSource> PrependIterator<TSource>(IEnumerable<TSource
+            }
+
+            protected override IEnumerable<TSource> GetEnumerable(int index)
+            {
+                switch (index)
+                {
+                    case 0: return _first;
+                    case 1: return _second;
+                    case 2: return _third;
+                    default: return null;
+                }
+            }
+        }
+
+        private sealed class ConcatNIterator<TSource> : ConcatIterator<TSource>
+        {
+            internal readonly IEnumerable<TSource>[] _sources;
+
+            internal ConcatNIterator(params IEnumerable<TSource>[] sources)
+            {
+                Debug.Assert(sources != null);
+                Debug.Assert(sources.All(s => s != null));
+                Debug.Assert(sources.Length > 3, "Should be using Concat2Iterator or Concat3Iterator");
+                _sources = sources;
+            }
+
+            public override Iterator<TSource> Clone()
+            {
+                return new ConcatNIterator<TSource>(_sources);
+            }
+
+            protected override IEnumerable<TSource> GetEnumerable(int index)
+            {
+                return index < _sources.Length ? _sources[index] : null;
+            }
+        }
+
+        private abstract class ConcatIterator<TSource> : Iterator<TSource>, IIListProvider<TSource>
+        {
+            private IEnumerator<TSource> _enumerator;
+
+            public override void Dispose()
+            {
+                if (_enumerator != null)
+                {
+                    _enumerator.Dispose();
+                    _enumerator = null;
+                }
+                base.Dispose();
+            }
+
+            protected abstract IEnumerable<TSource> GetEnumerable(int index);
+
+            public override bool MoveNext()
+            {
+                if (state == 1)
+                {
+                    _enumerator = GetEnumerable(0).GetEnumerator();
+                    state = 2;
+                }
+
+                if (state > 1)
+                {
+                    while (true)
+                    {
+                        if (_enumerator.MoveNext())
+                        {
+                            current = _enumerator.Current;
+                            return true;
+                        }
+
+                        Debug.Assert(state < int.MaxValue);
+                        IEnumerable<TSource> next = GetEnumerable(state++ - 1);
+                        if (next != null)
+                        {
+                            _enumerator.Dispose();
+                            _enumerator = next.GetEnumerator();
+                            continue;
+                        }
+
+                        Dispose();
+                        break;
+                    }
+                }
+
+                return false;
+            }
+
+            public TSource[] ToArray()
+            {
+                return EnumerableHelpers.ToArray(this);
```

Comment:
Likewise, `GetCount(true)` might let you preallocate, with this as the fall-back when the result is -1.